IPython/kernel/comm/comm.py

```diff
@@ -63,6 +66,10 @@ def __init__(self, target_name='', data=None, **kwargs):
     
     def _publish_msg(self, msg_type, data=None, metadata=None, buffers=None, **keys):
         """Helper for sending a comm message on IOPub"""
+        if threading.current_thread().name != 'MainThread' and IOLoop.initialized():
+            # make sure we never send on a zmq socket outside the main IOLoop thread
+            IOLoop.instance().add_callback(lambda : self._publish_msg(msg_type, data, metadata, buffers, **keys))
```

Comment:
@minrk is there a chance that the callback list will overflow if you are adding callbacks to it quickly?  In other words, does the IOLoop instance do something like this:

```python
def tick(self):
    callback = self.callbacks.pop()
    callback()
```

or does it clear the whole buffer each iteration?

```python
def tick(self):
    while len(self.callbacks) > 0:
        callback = self.callbacks.pop()
        callback()
```